homeassistant/components/kodi/config_flow.py

```diff
@@ -0,0 +1,264 @@
+
+        if user_input is not None:
+            self._host = user_input[CONF_HOST]
+            self._port = user_input[CONF_PORT]
+            self._ssl = user_input[CONF_SSL]
+            return await self.async_step_credentials()
+
+        return self.async_show_form(
+            step_id="host", data_schema=self._host_schema(), errors=errors
+        )
+
+    async def async_step_credentials(self, user_input=None):
+        """Handle username and password input."""
+        errors = {}
+        if user_input is not None:
+            self._username = user_input.get(CONF_USERNAME)
+            self._password = user_input.get(CONF_PASSWORD)
+            try:
+                await validate_http(self.hass, self._get_data())
+                return await self.async_step_ws_port()
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except CannotConnect:
+                if self._discovery_name:
+                    return self.async_abort(reason="cannot_connect")
+                return await self.async_step_host(errors={"base": "cannot_connect"})
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        return self.async_show_form(
+            step_id="credentials", data_schema=self._credentials_schema(), errors=errors
+        )
+
+    async def async_step_ws_port(self, user_input=None):
+        """Handle websocket port of discovered node."""
+        errors = {}
+        if user_input is not None:
+            self._ws_port = user_input.get(CONF_WS_PORT)
+            try:
+                await validate_ws(self.hass, self._get_data())
+                return self._create_entry()
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+        return self.async_show_form(
+            step_id="ws_port", data_schema=self._ws_port_schema(), errors=errors
+        )
+
+    async def async_step_import(self, data):
+        """Handle import from YAML."""
+        # We assume that the imported values work and just create the entry
+        return self.async_create_entry(title=data[CONF_NAME], data=data)
+
+    @callback
+    def _create_entry(self):
+        return self.async_create_entry(
+            title=self._name or self._host, data=self._get_data(),
+        )
+
+    @callback
+    def _get_data(self):
+        data = {
+            CONF_NAME: self._name,
+            CONF_HOST: self._host,
+            CONF_PORT: self._port,
+            CONF_WS_PORT: self._ws_port,
+            CONF_USERNAME: self._username,
+            CONF_PASSWORD: self._password,
+            CONF_SSL: self._ssl,
+            CONF_TIMEOUT: DEFAULT_TIMEOUT,
+        }
+        if self._uuid:
```

Comment:
thisseems extraneous when it would already be set as unique_id of the config entry?